File: preader/preader.go
package preader

import (
	"errors"
	"fmt"
	"io"
	"os"

	"golang.org/x/term"
)

type PassphraseReader interface {
	ReadPassphrase() (string, error)
}

func NewTerminal() PassphraseReader {
	return &terminalPassphraseReader{}
}

func NewCaching(upstream PassphraseReader) PassphraseReader {
	return &cachingPassphraseReader{Upstream: upstream}
}

func NewReader(reader io.Reader) PassphraseReader {
	return &readerPassphraseReader{reader: reader}
}

func NewConstant(passphrase string) PassphraseReader {
	return &constantPassphraseReader{passphrase: passphrase}
}

type constantPassphraseReader struct {
	passphrase string
}

func (r *constantPassphraseReader) ReadPassphrase() (string, error) {
	return r.passphrase, nil
}

type terminalPassphraseReader struct{}

func (r *terminalPassphraseReader) ReadPassphrase() (string, error) {
	fd := int(os.Stdin.Fd())
	if !term.IsTerminal(fd) {
		return "", errors.New("cannot read passphrase from terminal - stdin is not a terminal")
	}

	_, err := fmt.Fprint(os.Stderr, "Passphrase (saltybox): ")
	if err != nil {
		return "", err
	}
	phrase, err := term.ReadPassword(fd)
	if err != nil {
		return "", fmt.Errorf("failure reading passphrase: %w", err)
	}

	return string(phrase), nil
}

// cachingPassphraseReader will wrap a PassphraseReader by adding caching.
//
// This is useful to allow "at most once" semantics when reading the passphrase, while
// still lazily deferring the first invocation.
type cachingPassphraseReader struct {
	Upstream         PassphraseReader
	cachedPassphrase string
	cached           bool
}

func (r *cachingPassphraseReader) ReadPassphrase() (string, error) {
	if !r.cached {
		cached, err := r.Upstream.ReadPassphrase()
		if err != nil {
			return "", err
		}
		r.cachedPassphrase = cached
		r.cached = true
	}

	return r.cachedPassphrase, nil
}

type readerPassphraseReader struct {
	reader io.Reader
}

func (r *readerPassphraseReader) ReadPassphrase() (string, error) {
	data, err := io.ReadAll(r.reader)
	if err != nil {
		return "", fmt.Errorf("error reading passphrase: %w", err)
	}

	return string(data), nil
}

File: saltybox-rs/src/preader.rs
use std::io::{self, Read};

pub trait PassphraseReader {
    fn read_passphrase(&mut self) -> io::Result<String>;
}

impl<T: PassphraseReader + ?Sized> PassphraseReader for Box<T> {
    fn read_passphrase(&mut self) -> io::Result<String> {
        (**self).read_passphrase()
    }
}

pub struct TerminalPassphraseReader;

impl TerminalPassphraseReader {
    pub fn new() -> Self { Self }
}

impl PassphraseReader for TerminalPassphraseReader {
    fn read_passphrase(&mut self) -> io::Result<String> {
        let pass = rpassword::prompt_password("Passphrase (saltybox): ")
            .map_err(|e| io::Error::new(io::ErrorKind::Other, format!("failure reading passphrase: {}", e)))?;
        Ok(pass)
    }
}

pub struct ReaderPassphraseReader<R: Read> {
    reader: R,
}

impl<R: Read> ReaderPassphraseReader<R> {
    pub fn new(reader: R) -> Self { Self { reader } }
}

impl<R: Read> PassphraseReader for ReaderPassphraseReader<R> {
    fn read_passphrase(&mut self) -> io::Result<String> {
        let mut data = String::new();
        self.reader.read_to_string(&mut data)?;
        Ok(data)
    }
}

pub struct ConstantPassphraseReader {
    passphrase: String,
}

impl ConstantPassphraseReader {
    pub fn new(passphrase: String) -> Self { Self { passphrase } }
}

impl PassphraseReader for ConstantPassphraseReader {
    fn read_passphrase(&mut self) -> io::Result<String> {
        Ok(self.passphrase.clone())
    }
}

pub struct CachingPassphraseReader<P: PassphraseReader> {
    upstream: P,
    cached: Option<String>,
}

impl<P: PassphraseReader> CachingPassphraseReader<P> {
    pub fn new(upstream: P) -> Self { Self { upstream, cached: None } }
}

impl<P: PassphraseReader> PassphraseReader for CachingPassphraseReader<P> {
    fn read_passphrase(&mut self) -> io::Result<String> {
        if let Some(cached) = &self.cached { return Ok(cached.clone()); }
        let phrase = self.upstream.read_passphrase()?;
        self.cached = Some(phrase.clone());
        Ok(phrase)
    }
}

